File: docs/codec.js
// claude-snap codec — JS port of the Python unpack/pack logic.
// Pure module, no DOM, no I/O. Roundtrip-compatible with the Python reference.
// MIT.

export const SNAP_HEADER = 'snap_header';
export const SNAP_FOOTER = 'snap_footer';
export const SNAP_EVENT = 'snap_event';
export const SNAP_REF = 'snap_ref';
export const SNAP_DANGLING_REF = 'snap_dangling_ref';

/**
 * Parse a JSONL string into an array of objects.
 * Bad lines are skipped silently to match the Python reader's behavior.
 * @param {string} text
 * @returns {Array<object>}
 */
export function readJsonl(text) {
  const out = [];
  if (typeof text !== 'string') return out;
  const lines = text.split('\n');
  for (const raw of lines) {
    const line = raw.trim();
    if (!line) continue;
    try {
      out.push(JSON.parse(line));
    } catch (_e) {
      // Match Python codec: silently skip malformed lines.
    }
  }
  return out;
}

/**
 * Serialize an array of objects as JSONL.
 * @param {Array<object>} records
 * @returns {string}
 */
export function writeJsonl(records) {
  return records.map(r => JSON.stringify(r)).join('\n') + '\n';
}

/**
 * Detect whether a parsed-JSONL list is a packed snap stream.
 * @param {Array<object>} records
 */
export function isPacked(records) {
  if (!Array.isArray(records) || records.length === 0) return false;
  const head = records[0];
  return head && typeof head === 'object' && head.type === SNAP_HEADER;
}

/**
 * Resolve refs back to full payloads. Mirrors codec.py:unpack().
 *
 * Property: for any list of canonical Events `evs`,
 *   unpack(pack(evs)) === [ev.payload for ev in evs]
 * (modulo header/footer metadata).
 *
 * @param {Array<object>} packed - parsed snap_* records
 * @returns {Array<object>} restored raw JSONL events in original order
 */
export function unpack(packed) {
  const byHash = new Map();
  const out = [];
  if (!Array.isArray(packed)) return out;

  for (const x of packed) {
    if (!x || typeof x !== 'object') continue;
    const t = x.type;

    if (t === SNAP_HEADER || t === SNAP_FOOTER) continue;

    if (t === SNAP_EVENT) {
      byHash.set(x.content_hash, x.payload);
      out.push(x.payload);
      continue;
    }

    if (t === SNAP_REF) {
      const refTo = x.ref_to;
      const payload = byHash.get(refTo);
      if (payload === undefined) {
        out.push({
          type: SNAP_DANGLING_REF,
          ref_to: refTo,
          seq: x.seq,
        });
        continue;
      }

      // Deep-copy via JSON round-trip — payloads are JSON-safe.
      const restored = JSON.parse(JSON.stringify(payload));

      // Patch the inner tool_use_id so tool_use ↔ tool_result linkage stays correct.
      const newToolId = x.tool_id;
      if (newToolId && restored && typeof restored === 'object') {
        const msg = restored.message;
        if (msg && typeof msg === 'object') {
          const content = msg.content;
          if (Array.isArray(content)) {
            for (const b of content) {
              if (b && typeof b === 'object' && b.type === 'tool_result') {
                b.tool_use_id = newToolId;
              }
            }
          }
        }
      }

      // Overlay per-event metadata (uuid / parentUuid / timestamp / ...).
      const patch = x.patch;
      if (patch && typeof patch === 'object' && restored && typeof restored === 'object') {
        Object.assign(restored, patch);
      }

      out.push(restored);
    }
  }

  return out;
}

/**
 * Pull a flat list of {role, content} chat-style messages from a list of
 * raw Claude Code JSONL events, suitable for sending to Anthropic Messages API.
 *
 * - Skips meta/system/sidechain events.
 * - Concatenates assistant text blocks; drops thinking blocks (the API doesn't
 *   accept extended-thinking blocks as input from the messages array).
 * - Concatenates user text content; tool_results become text summaries.
 *
 * @param {Array<object>} events  raw events (unpacked or original)
 * @returns {Array<{role: string, content: string}>}
 */
export function eventsToMessages(events) {
  const msgs = [];
  if (!Array.isArray(events)) return msgs;

  for (const ev of events) {
    if (!ev || typeof ev !== 'object') continue;
    if (ev.isSidechain) continue;

    const t = ev.type;
    if (t !== 'user' && t !== 'assistant') continue;

    const inner = ev.message;
    if (!inner || typeof inner !== 'object') continue;

    const content = inner.content;
    let text = '';

    if (typeof content === 'string') {
      text = content;
    } else if (Array.isArray(content)) {
      const parts = [];
      for (const b of content) {
        if (!b || typeof b !== 'object') continue;
        switch (b.type) {
          case 'text':
            if (typeof b.text === 'string') parts.push(b.text);
            break;
          case 'thinking':
            // Drop thinking blocks — they don't replay through the API.
            break;
          case 'tool_use': {
            const name = b.name || 'tool';
            const input = b.input ? JSON.stringify(b.input) : '';
            parts.push(`[tool_use ${name}${input ? ` ${input}` : ''}]`);
            break;
          }
          case 'tool_result': {
            const inner2 = b.content;
            if (typeof inner2 === 'string') {
              parts.push(`[tool_result]\n${inner2}`);
            } else if (Array.isArray(inner2)) {
              const txt = inner2
                .filter(x => x && typeof x === 'object' && typeof x.text === 'string')
                .map(x => x.text)
                .join('\n');
              if (txt) parts.push(`[tool_result]\n${txt}`);
            }
            break;
          }
          default:
            break;
        }
      }
      text = parts.join('\n\n').trim();
    }

    if (!text) continue;

    const role = (t === 'user') ? 'user' : 'assistant';
    // Collapse adjacent same-role messages for API compatibility.
    if (msgs.length > 0 && msgs[msgs.length - 1].role === role) {
      msgs[msgs.length - 1].content += '\n\n' + text;
    } else {
      msgs.push({ role, content: text });
    }
  }

  return msgs;
}

/**
 * Surface stats on a packed stream — port of codec.py:stats().
 * @param {Array<object>} packed
 */
export function stats(packed) {
  let events = 0;
  let refs = 0;
  const eventBytesByHash = new Map();
  if (!Array.isArray(packed)) {
    return { events, refs, events_plus_refs: 0, bytes_unpacked: 0, bytes_packed: 0, compression_ratio: 1 };
  }
  for (const x of packed) {
    if (x && x.type === SNAP_EVENT) {
      eventBytesByHash.set(x.content_hash, JSON.stringify(x).length);
    }
  }
  let fullBytes = 0;
  let packedBytes = 0;
  for (const x of packed) {
    if (!x || typeof x !== 'object') continue;
    if (x.type === SNAP_EVENT) {
      events += 1;
      const b = JSON.stringify(x).length;
      fullBytes += b;
      packedBytes += b;
    } else if (x.type === SNAP_REF) {
      refs += 1;
      fullBytes += eventBytesByHash.get(x.ref_to) || 0;
      packedBytes += JSON.stringify(x).length;
    }
  }
  const ratio = fullBytes / Math.max(packedBytes, 1);
  return {
    events,
    refs,
    events_plus_refs: events + refs,
    bytes_unpacked: fullBytes,
    bytes_packed: packedBytes,
    compression_ratio: Math.round(ratio * 100) / 100,
  };
}

File: claude_snap/codec.py
"""
The claude-snap codec.

Heuristics, in priority order:
  1. User and assistant messages: kept verbatim, always.
  2. Edit / Write tool calls and their results: kept verbatim, always.
  3. First Read of a given file: kept verbatim.
  4. Re-Read of a file with identical content AND no Edit/Write to that path
     between the two reads: replace the second read's tool_result with a
     snap_ref pointing to the first.
  5. Identical Bash tool_results (content-hash match): ref the duplicate.
  6. Anything else (TodoWrite, MultiEdit, Glob, Grep, custom MCP tools):
     pass through verbatim.

Correctness condition: unpack(pack(events)) == events byte-for-byte,
modulo the header/footer metadata. When resolving a ref on unpack, the
restored payload's tool_use_id is patched to the ref event's id so the
tool_use ↔ tool_result linkage stays correct, and any captured per-event
metadata (uuid, parentUuid, timestamp, ...) is overlaid on the restored
payload.
"""

from __future__ import annotations
import copy
import json
from typing import Iterator

from .schema import (
    Event, REF, USER_MSG, ASSISTANT_MSG, TOOL_USE, TOOL_RESULT, META,
    normalize, is_mutating, READ_LIKE, SHELL, NETWORK,
)


CLAUDE_SNAP_VERSION = "0.3.0"

# Tool families whose results are safe to dedup. Mutating tools
# (Edit/Write/MultiEdit/NotebookEdit) and meta tools (TodoWrite/Task) are
# excluded because their results carry operation-specific data that varies
# per call even when the visible message is the same.
SAFE_TO_DEDUP_FAMILIES = READ_LIKE | SHELL | NETWORK

# Per-event metadata fields that vary between otherwise-identical events.
# Captured on the ref so unpack can restore byte-identical payloads.
PER_EVENT_METADATA_FIELDS = (
    "uuid", "parentUuid", "timestamp", "sessionId", "requestId",
    "sourceToolAssistantUUID", "promptId", "messageId", "leafUuid",
    "lastPrompt",
)


def _read_jsonl(path: str) -> Iterator[dict]:
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                yield json.loads(line)
            except json.JSONDecodeError:
                continue


def parse(jsonl_path: str) -> list:
    """Read a Claude Code JSONL file into a list of canonical Events."""
    return [normalize(raw, i) for i, raw in enumerate(_read_jsonl(jsonl_path))]


def pack(events: list) -> list:
    """Apply the dedup heuristics. Returns JSON-serializable dicts."""
    out: list = [{
        "type": "snap_header",
        "version": CLAUDE_SNAP_VERSION,
        "original_event_count": len(events),
    }]

    last_read_hash: dict = {}        # (tool, path) -> hash
    last_mutation_seq: dict = {}     # path -> seq
    seen_result_hash: dict = {}      # hash -> seq
    tool_use_index: dict = {}        # tool_id -> Event

    for ev in events:
        if ev.kind == TOOL_USE and is_mutating(ev) and ev.target_path:
            last_mutation_seq[ev.target_path] = ev.seq

        if ev.kind in (USER_MSG, ASSISTANT_MSG, META):
            out.append(_emit(ev))
            continue

        if ev.kind == TOOL_USE:
            out.append(_emit(ev))
            if ev.tool_id:
                tool_use_index[ev.tool_id] = ev
            continue

        if ev.kind == TOOL_RESULT:
            matched_use = tool_use_index.get(ev.tool_id)

            # Only dedup tool_results from safe (non-mutating) families.
            # Edit/Write results carry operation-specific toolUseResult data
            # (oldString/newString/structuredPatch) and must never be ref'd.
            safe = (matched_use is not None
                    and matched_use.tool_name in SAFE_TO_DEDUP_FAMILIES)

            if safe:
                # Heuristic 4: re-read of unchanged file
                if matched_use.tool_name in READ_LIKE \
                        and matched_use.target_path:
                    key = (matched_use.tool_name, matched_use.target_path)
                    prior_hash = last_read_hash.get(key)
                    last_mut = last_mutation_seq.get(
                        matched_use.target_path, -1)

                    if prior_hash == ev.content_hash and last_mut < ev.seq:
                        out.append(_emit_ref(
                            ev, prior_hash,
                            reason="unchanged_read",
                            path=matched_use.target_path,
                        ))
                        continue
                    last_read_hash[key] = ev.content_hash

                # Heuristic 5: duplicate tool_result content (Bash etc.)
                if ev.content_hash in seen_result_hash:
                    out.append(_emit_ref(
                        ev, ev.content_hash,
                        reason="duplicate_result",
                    ))
                    continue
                seen_result_hash[ev.content_hash] = ev.seq

            out.append(_emit(ev))
            continue

        out.append(_emit(ev))

    refs = sum(1 for x in out if x.get("type") == "snap_ref")
    out.append({
        "type": "snap_footer",
        "events_in": len(events),
        "events_out": len(out) - 2,
        "refs_introduced": refs,
    })
    return out


def _emit(ev: Event) -> dict:
    """Serialize an Event to a snap JSONL line. Drop None fields."""
    rec = {
        "type": "snap_event",
        "kind": ev.kind,
        "seq": ev.seq,
        "content_hash": ev.content_hash,
        "payload": ev.payload,
    }
    if ev.tool_name is not None:
        rec["tool_name"] = ev.tool_name
    if ev.tool_id is not None:
        rec["tool_id"] = ev.tool_id
    if ev.target_path is not None:
        rec["target_path"] = ev.target_path
    return rec


def _emit_ref(ev: Event, ref_to: str, reason: str, path: str = None) -> dict:
    out = {
        "type": "snap_ref",
        "kind": REF,
        "seq": ev.seq,
        "ref_to": ref_to,
        "reason": reason,
        "tool_id": ev.tool_id,
    }
    if path:
        out["target_path"] = path
    # Capture per-event metadata so unpack restores byte-identical payloads.
    patch = {k: ev.payload[k] for k in PER_EVENT_METADATA_FIELDS
             if isinstance(ev.payload, dict) and k in ev.payload}
    if patch:
        out["patch"] = patch
    return out


def unpack(packed: list) -> list:
    """
    Resolve refs back to full payloads. Returns the original raw JSONL
    dicts in original order.

    Property: for any Claude Code session JSONL `f`,
        unpack(pack(parse(f))) == [ev.payload for ev in parse(f)]
    """
    by_hash: dict = {}
    out: list = []

    for x in packed:
        t = x.get("type")
        if t in ("snap_header", "snap_footer"):
            continue
        if t == "snap_event":
            by_hash[x["content_hash"]] = x["payload"]
            out.append(x["payload"])
            continue
        if t == "snap_ref":
            ref_to = x.get("ref_to")
            payload = by_hash.get(ref_to)
            if payload is None:
                out.append({
                    "type": "snap_dangling_ref",
                    "ref_to": ref_to,
                    "seq": x.get("seq"),
                })
                continue
            restored = copy.deepcopy(payload)
            new_tool_id = x.get("tool_id")
            if new_tool_id and isinstance(restored, dict):
                msg = restored.get("message", {})
                content = msg.get("content") if isinstance(msg, dict) else None
                if isinstance(content, list):
                    for b in content:
                        if isinstance(b, dict) and b.get("type") == "tool_result":
                            b["tool_use_id"] = new_tool_id
            patch = x.get("patch")
            if isinstance(patch, dict) and isinstance(restored, dict):
                restored.update(patch)
            out.append(restored)
            continue

    return out


def write_jsonl(records: list, path: str) -> None:
    with open(path, "w", encoding="utf-8") as f:
        for r in records:
            f.write(json.dumps(r, ensure_ascii=False) + "\n")


def stats(packed: list) -> dict:
    """Compression stats from a packed stream."""
    events = sum(1 for x in packed if x.get("type") == "snap_event")
    refs = sum(1 for x in packed if x.get("type") == "snap_ref")

    event_bytes_by_hash: dict = {}
    for x in packed:
        if x.get("type") == "snap_event":
            event_bytes_by_hash[x["content_hash"]] = len(json.dumps(x))

    full_bytes = 0
    packed_bytes = 0
    for x in packed:
        t = x.get("type")
        if t == "snap_event":
            b = len(json.dumps(x))
            full_bytes += b
            packed_bytes += b
        elif t == "snap_ref":
            full_bytes += event_bytes_by_hash.get(x.get("ref_to"), 0)
            packed_bytes += len(json.dumps(x))

    return {
        "events": events,
        "refs": refs,
        "events_plus_refs": events + refs,
        "bytes_unpacked": full_bytes,
        "bytes_packed": packed_bytes,
        "compression_ratio": round(full_bytes / max(packed_bytes, 1), 2),
    }
